Implement "unused pattern bindings" p2022 - parsing

burakemir opened the pull request:
This implements proposal #2022, with changes from #3763 and leads answers on #6448

Detection of unusedness is happening in ~~`toolchain/check/dataflow_analysis.cpp`~~ next PR. See [here](https://github.com/burakemir/carbon-lang/tree/unused_pattern_bindings_p2022_impl_part2) for preview.

~~All test cases with unused bindings were updated in order to avoid polluting test output.~~

burakemir commented:
One detail: when a variable is only used to access some type, and not used otherwise, it still counts as unused.

See toolchain/check/testdata/class/nested_name.carbon

I feel it may be better to keep detection of unusedness in types for a different PR.

danakj commented:
> This implements proposal #2022, with changes from #3763 and leads answers on #6448

Could I also trouble you to make sure the docs are up to date with all the changes, maybe in a separate PR if you like?

https://github.com/carbon-language/carbon-lang/blob/103c49a763f03f6ddecbde15775d6deb2b924a8f/docs/design/pattern_matching.md#unused-bindings

danakj commented on toolchain/parse/testdata/var/fail_unused.carbon:
```diff
@@ -0,0 +1,43 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/parse/testdata/var/fail_unused.carbon
+
+var unused x: i32;
```
Is this supposed to have an error?

burakemir replied:
No, it is not. Moved to var.carbon

danakj commented on toolchain/parse/testdata/var/fail_unused.carbon:
```diff
@@ -0,0 +1,43 @@
+// CHECK:STDERR:     ^~~
+// CHECK:STDERR:
+var ref unused y: i32;
+var unused _: i32;
```
Is this supposed to have an error?

We want to avoid mixing failures and passes in the same file split, so that if one flips to the other, we get a FAIL from the overall test run.

This set of tests here seems both a little oddly specific but also beyond the scope of `unused` so I am not sure what to make of it.

It's not testing putting `unused` in all possible positions, but it is testing putting `var` and `ref` together. So maybe we need more tests here, to verify parsing unused in all the positions.

I would suggest using function parameters, cuz you can't even write `ref` on a variable decl.

burakemir replied:
I have added "unused" to a few more places now. I am hesitating to add a whole "unused" directory, as once diagnostics are added, all test cases that have unused bindings (= almost all of them) will require unused, in various places.

This is why testing the specific thing that is not permitted (ref unused) seems to be the right choice to me.

danakj replied:
I would just rename this (disk) file to `unused.carbon` and then add file splits inside it for each test case. I don't know that a directory is needed at this point.

I am getting a feeling you are not familiar with file splits, do you know what I mean by that?

burakemir replied:
I noticed those! Will do.

danakj commented on toolchain/parse/handle_var.cpp:
```diff
@@ -156,4 +156,23 @@ auto HandleFinishVariablePattern(Context& context) -> void {
   }
 }
 
+auto HandleUnusedPattern(Context& context) -> void {
```
Should these go in a handle_unused.cpp? They don't seem specific to `var` at all

burakemir replied:
Done

danakj commented on toolchain/parse/state.def:
```diff
@@ -1085,6 +1085,21 @@ CARBON_PARSE_STATE(VariablePattern)
+// ^~~~~~
+//   1. Pattern
+//   2. FinishUnusedPattern
+CARBON_PARSE_STATE(UnusedPattern)
```
Nothing above currently references `UnusedPattern` which doesn't look right. See for example on L1033 where `VariablePattern` is referenced. We'll need to explain/enumerate all the places `UnusedPattern` can show up.

Also in https://github.com/carbon-language/carbon-lang/issues/6448#issuecomment-3599303524, @zygoloid referred to `unused` as a "pattern operator" and maybe it would make sense to use that name here as well, rather than calling it a `Pattern` which sounds incorrect to me?

It probably makes sense to move these up beside `Pattern` rather than down here - this file is not alphabetical.

burakemir replied:
Mentioned this in "Pattern" above now, next to where "VariablePattern" is mentioned.

"unused" and "var" in patterns are merely operators in the sense that they are going to be inner nodes in the pattern syntax tree. I'd prefer to keep it consistent with VariablePattern.

Since they are so similar, I find it useful that these are located right next to VariablePattern.

danakj reviewed (commented):
I have only reviewed the parse stuff so far.

Can we move the check/ changes into another PR, so that we can review them in isolation? https://github.com/carbon-language/carbon-lang/pull/6255/files#diff-c7f2de42def7e3e0072ee66501047e5e72e0578acae95f5e6a4037b1b4648b2b is an example of another PR that does similar.

Then we'd also have the option of letting check just ignore `unused` and we could insert it into tests as a PR on its own, then follow that with the changes to check/ to enforce it, along with any test changes that result from enforcement specifically?

burakemir committed:
Unused binding marker and diagnostics, part 1: parsing

Modifies the parser to accept "unused" as a pattern operator.
"unused" is implemented in a way that is consistent with "var".
Adds test cases, in toolchain/parse/testdata/var/unused.carbon.

Handling the multiple diagnostics in the check-phases is left as TODO
here.

danakj commented:
Meta note: It is preferable to avoid pushing rebases once a review has started as it tends to orphan comments from previous review rounds.

burakemir commented:
Should I close and open a new one?

I don't see how I could have split up the series without changing the commit id.

danakj commented:
> Should I close and open a new one?
> 
> I don't see how I could have split up the series without changing the commit id.

It's fine, I'll work with it. To avoid a rebase would have just meant reverting unwanted changes in the PR with a commit.

burakemir committed:
Adds diagnostic for "nested unused" error.

It is an error to mark patterns "unused" multiple times
in a nested manner. This is the same as for "var" and is
thus handled in the parse phase by an extra bit of pattern state.

burakemir committed:
Update docs with unused-pattern marker

burakemir committed:
Add TODO test cases for upcoming check-diagnostics.

burakemir commented:
I have added the "parse" diagnostics, consolidated test cases and made an update to the documentation. PTAL.

Not sure what you have in mind for the hundreds of test cases that have unused bindings and which legitimately trigger warnings - but we can keep that discussion for the next PR (part 2).

danakj commented on docs/design/pattern_matching.md:
```diff
@@ -345,6 +341,29 @@ _pattern_ `=` _expression_ `;`.
 
+### `unused`
+
+An `unused` prefix indicate that names are visible for name lookup but uses are
```
```suggestion
An `unused` prefix indicates that names are visible for name lookup but uses are
```

burakemir replied:
Done

danakj commented on docs/design/pattern_matching.md:
```diff
@@ -345,6 +341,29 @@ _pattern_ `=` _expression_ `;`.
+
+An `unused` pattern cannot be nested within another `unused` pattern.
+
+As specified in [#3763](/proposals/p3763.md), the `unused` marker may only
```
Should we discuss at all how you can place unused outside a tuple binding or inside it, and on either side of the `var` and what each means?

burakemir replied:
I have changed to "names in its nested pattern", and added a few sentences. How does that look?

burakemir replied:
Reworded a bit further.

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,78 @@
+}
+
+fn Parameters(p: i32) {
+}
```
These all look like pass cases. Can they move out of a `fail_` file split?

burakemir replied:
Yes, these are warnings, not errors.
Moving up.

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,78 @@
+  // CHECK:STDERR:       ^~~~~~~~~~~~~
+  // CHECK:STDERR:
+  var unused y: i32 = x;
+}
```
I think there's two different test cases here:

1. A test that fails now but will pass later:
  fail_todo_...
  ```
  // error: semantics TODO: `unused`
  var unused y: i32 = x;
  ```
2. A test that fails now and should fail later
  fail_...
  ```
   var unused y: i32  = x;
  // TODO: Should be an error to use an `unused` name.
   y += 1;
   ```

burakemir replied:
I have added the // TODO lines to indicate what should happen.

See here for a sneak peak at actual warnings, errors:
https://github.com/burakemir/carbon-lang/blob/unused_pattern_bindings_p2022_impl_part2/toolchain/check/testdata/dataflow/unused.carbon

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,78 @@
+  var unused y: i32 = x;
+}
+
+fn RefParamUnused(ref y: i32) {
```
Similarly there are at least two cases here:

1. `unused ref y` is not used (fail_todo_...)
2. `unused ref y` is used (fail_...)
3. `ref y` is not used (todo_fail...)
4.  `ref y` is used (pass)

burakemir replied:
I added 1. and 2., cases 3. and 4 are covered further up (RefParam and RefParamUnused).

When a binding is unused without being marked "unused", that is a warning, not an error.

danakj commented on toolchain/parse/handle_var.cpp:
```diff
@@ -25,7 +25,8 @@ static auto HandleVar(Context& context, StateKind finish_state_kind,
 
-  context.PushStateForPattern(StateKind::Pattern, /*in_var_pattern=*/true);
+  context.PushStateForPattern(StateKind::Pattern, /*in_var_pattern=*/true,
+                              /*in_unused_pattern=*/false);
```
Will this do the wrong thing for `unused var ...` ? Do we have tests for that? Something like `unused var unused x: i32` ?

burakemir replied:
Since unused does not start a declaration, there are no such tests.
Added a fail_ test to parse/testdata/var/unused.carbon.

danakj replied:
We can see unused before var in some cases though, right? like `case unused (var unused a: i32, b: i32)`? Are there others like that, any we can test? Maybe they don't land here in the code?

burakemir replied:
Yes, there is a difference between:
* var as decl (via StatementLoop, DeclScopeLoop, VarAsRegular/ VarAsReturned), so this is really at the beginning cannot have "unused" state - unless we really want "unused" to start declaration.
* var as pattern (VariablePattern)

So within a pattern the order of unused and var does not matter. I added a match statement test case to "dataflow/unused.carbon", it shows that parsing works, although match is not implemented in check.

danakj replied:
Ok, thanks. If it would make sense, the test could be in parse instead, but up to you.

I guess I didn't see `unused var` for a parameter in the parse tests. I will comment there.

burakemir replied:
Added there.

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,412 @@
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/parse/testdata/var/unused.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/parse/testdata/var/unused.carbon
+
```
I am guessing that `unused` should always be an error in a `returned` context (https://github.com/carbon-language/carbon-lang/blob/154e4012c4cf875a11caf9137330ef2309fd7437/toolchain/parse/testdata/return/returned_var.carbon#L12)

like `return unused x: i32` would be an oxymoron cuz `return var` later is going to return the x?

Not sure if that should be a parse or a check error, but it'd be nice to be in parse if we can.

burakemir replied:
Good point. I think this is easier to handle in check, where also the other "UnusedButUsed" diagnostics are implemented. I believe it should already work.

Adding "returned var unused x: i32" as a test case to the dataflow/ununused.carbon file.

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,412 @@
+// CHECK:STDERR: var ref unused y: i32;
+// CHECK:STDERR:     ^~~
+// CHECK:STDERR:
+var ref unused y: i32;
```
This isn't really testing unused, it ends up finding an error between var and ref. Maybe we want something like this?
```
fn F(ref unused y: i32)
```

burakemir replied:
Done

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,412 @@
+// CHECK:STDERR: var ref unused y: i32;
+// CHECK:STDERR:     ^~~
+// CHECK:STDERR:
+var ref unused y: i32;
```
This is the same as the previous test?

burakemir replied:
Removed.

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,412 @@
+// CHECK:STDERR: fn I(x: (), unused var ref y: ());
+// CHECK:STDERR:                        ^~~
+// CHECK:STDERR:
+fn I(x: (), unused var ref y: ());
```
These are also testing var-ref interactions. I would have expected us to not put `var` in these so we can test unused against ref interactions?

burakemir replied:
Removed the redundant ones.

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,412 @@
+// CHECK:STDERR: var unused (unused y: i32) = (0,);
+// CHECK:STDERR:             ^~~~~~
+// CHECK:STDERR:
+var unused (unused y: i32) = (0,);
```
Looks like we're missing `unused var ... unused`?

burakemir replied:
`unused` cannot start a declaration though.

burakemir committed:
Apply reviewer comments

burakemir committed:
One more test case.

burakemir commented:
BTW, I have broken out uninteresting updates to toolchain/lower/testdata test cases into #6479

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+// --- todo_ensure_no_warning.carbon
 
-library "no_unused_error";
+library "no_unused_warning";
```
```suggestion
library "[[@TEST_NAME]]";
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
 }
 
-// --- fail_unused.carbon
+// --- fail_todo_ensure_no_warning.carbon
```
(Without renaming one of them, this ends up having the same library name as the above test: the fail_ does not contribute to the name).

```suggestion
// --- fail_todo_unused_ref_param.carbon
library "[[@TEST_NAME]]";
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+
+// --- todo_ensure_warning.carbon
+
+library "unused_warning";
```
```suggestion
library "[[@TEST_NAME]]";
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: ensure no warning
+}
+
+// --- todo_ensure_warning.carbon
```
We want this test to fail (so it's todo_fail_)

```suggestion
// --- todo_fail_unused_names.carbon
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
 // TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/dataflow/unused.carbon
 
-// --- no_unused_error.carbon
+// --- todo_ensure_no_warning.carbon
```
```suggestion
// --- todo_used_names.carbon
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+// --- fail_todo_ensure_error.carbon
+// CHECK:STDERR: fail_todo_ensure_error.carbon: error: `Main//default` previously provided by `fail_todo_ensure_no_warning.carbon` [DuplicateMainApi]
+// CHECK:STDERR:
+
```
```suggestion
// --- fail_todo_read_from_param_unused_ref.carbon
library "[[@TEST_NAME]]";
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: ensure error, marked "unused" but used.
+}
+
+fn ReturnVarUnused() -> i32 {
```
```suggestion
// --- todo_fail_unused_returned_var.carbon
library "[[@TEST_NAME]]";

fn F() -> i32 {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  let y: i32 = 1;
+  var _: i32 = y;
+  // TODO: ensure no warning.
 }
```
```suggestion
fn ReturnVar() -> i32 {
  var x: i32 = 0;
  x = 1;
  return x;
   // TODO: ensure no warning.
 }

fn ReturnLet() -> i32 {
  let y: i32 = 1;
  return y;
  // TODO: ensure no warning.
}
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+
+library "unused_warning";
 
 fn LetUnused() {
```
```suggestion
// --- todo_fail_local_let_named_not_used.carbon
library "[[@TEST_NAME]]";

fn F() {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: warning, x is unused.
 }
 
 fn VarUnused() {
```
```suggestion
// --- todo_fail_local_var_named_not_used.carbon
library "[[@TEST_NAME]]";

fn F() {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: warning, a is unused.
 }
 
 fn Shadowing() {
```
```suggestion
// --- todo_fail_local_var_shadowed_not_used.carbon
library "[[@TEST_NAME]]";

fn F() {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: warning, x is unused.
 }
 
 fn Parameters(p: i32) {
```
```suggestion
// --- todo_fail_param_let_named_not_used.carbon
library "[[@TEST_NAME]]";

fn F(p: i32) {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -10,69 +10,113 @@
+  // TODO: warning, p is unused.
+}
+
+fn RefParamUnused(ref y: i32) {
```
```suggestion
// --- todo_fail_param_ref_named_not_used.carbon
library "[[@TEST_NAME]]";

fn F(ref y: i32) {
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,122 @@
+
+// --- todo_ensure_no_warning.carbon
+
+library "no_unused_warning";
```
```suggestion
library "[[@TEST_NAME]]";
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,122 @@
+// --- todo_ensure_warning.carbon
+
+library "unused_warning";
+
```
```suggestion
```

danakj commented on toolchain/check/testdata/dataflow/unused.carbon:
```diff
@@ -0,0 +1,122 @@
+}
+
+fn Underscore(x: i32) {
+  var _: i32 = x;
```
```suggestion
// --- todo_fail_read_from_unused_local_var.carbon
library "[[@TEST_NAME]]";
	
fn F(x: i32) {
  var _: i32 = x;
```

danakj reviewed (commented):
It's nice to split test cases into separate file splits, especially when we get into failing tests. Ideally we keep failures one to a file, so that we get the largest signal from `file_test` when a failure stops happening by accident. I've added suggestions to split up the tests into a lot of splits as a result. Also introducing the `library [[@TEST_NAME]]` to each one, as this is how we name libraries in file_test.

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Update toolchain/check/testdata/dataflow/unused.carbon

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Add fail_ prefix, and todo test case for match expressions

burakemir committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,333 @@
+// CHECK:STDERR:
+var unused unused x: i32 = 0;
+
+// CHECK:STDERR: fail_nested.carbon:[[@LINE+4]]:13: error: `unused` nested within another `unused` [NestedUnused]
```
```suggestion
// --- fail_nested_parens.carbon

// CHECK:STDERR: fail_nested_parens.carbon:[[@LINE+4]]:13: error: `unused` nested within another `unused` [NestedUnused]
```

(splitting failures into individual splits)

danakj commented on toolchain/parse/testdata/var/unused.carbon:
```diff
@@ -0,0 +1,333 @@
+let unused y: i32;
+let unused ref w: i32 = 0;
+
+fn F(unused x: i32, var unused b: i32, unused ref c: i32) {
```
can we also test `unused var` as a parameter?

burakemir replied:
Done

danakj reviewed (approved):
LGTM w/ a few more straggler comments, thanks

burakemir committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

burakemir committed:
Add unused var and match parse tests

burakemir reviewed (commented):
Thanks for the review!

burakemir committed:
Fix test, "unused var" is ok as parameter

burakemir commented:
As discussed, I have done a pull --rebase from trunk,, updated that one test.
Please merge.
